### PR TITLE
Viimeisimmät kurssit refaktorointi

### LINKED ISSUE #6620: Viimeisimmät kurssit refaktorointi
Muikun etusivulla oleva "Vikmeisimmät kurssit/jatka opintoja" tarvitsee uudelleen suunnittelun.

2 uutta endpointtia tarvitaan.

Ajatus on että kun opiskelija menee kurssin materiaaleihin, jossa on aktiivisena jäsenenä, niin tällöin tallennetaan käyttäjäkohtainen tieto asiasta kuten nytkin. Luultavasti riittää jos userId:n ja workspaceId:n ottaa endpoint vastaan.

Kun taas haetaan viimeisimmät kurssit etusivulla niin bäkkäri voisi toimittaa alustavasti seuraavanlaisen json:in per työtila:

{
  workspaceName: "",
  materialName: "",
  workspaceEvaluationStatus: "",
  assigmentsDone: "",
  assigmentsTotal: "",
  exercisesDone: "",
  exercisesTotal: "",
}

Bäkkärin olisi tarkoitus huolehtia että listasta poistuu automaattisesti ne työtilat joissa opiskelija ei ole aktiivisena jäsenenä. Suoritetut kurssit näkyisi mukana jos opiskelijaa ei ole arkistoitu kurssilta vielä. Tossa esimerkissä olevilla assignment ja exercise key:t ois mukana jotta voidaan näyttää tarkempaa dataa halutessaan.

### PR DESCRIPTION
**Pull request by saaralepisto**
closes #6620

### CODE REVIEW

File: muikku-core/src/main/java/fi/otavanopisto/muikku/users/WorkspaceUserEntityController.java [hunk 1]
```diff
@@ -205,6 +213,44 @@ public WorkspaceUserEntity updateIdentifier(WorkspaceUserEntity workspaceUserEnt
...
+          for (LastWorkspace lastWorkspace : lastWorkspaceList) {
+            if (lastWorkspace.getWorkspaceId().equals(workspaceUserEntity.getWorkspaceEntity().getId())) {
+              continue;
+            } else {
+              newList.add(lastWorkspace);
+            }
+          }
+        }
+        try {
+          if (newList.isEmpty()) {
```
Comment (pkukkon) [inline muikku-core/src/main/java/fi/otavanopisto/muikku/users/WorkspaceUserEntityController.java:243]: Shouldn't be needed. Front-end can most likely cope with `last-workspace` either being `null` or an empty array `[]`

**Review by pkukkon — CHANGES_REQUESTED** [review]

File: muikku-core/src/main/java/fi/otavanopisto/muikku/users/WorkspaceUserEntityController.java [hunk 2]
```diff
@@ -205,6 +215,38 @@ public WorkspaceUserEntity updateIdentifier(WorkspaceUserEntity workspaceUserEnt
   }
 
   public WorkspaceUserEntity updateActive(WorkspaceUserEntity workspaceUserEntity, Boolean active) {
+    if (Boolean.FALSE.equals(active)) {
+      // #6620: Remove past workspaces from the list of student's last workspaces
+      UserEntity userEntity = userEntityController.findUserEntityByDataSourceAndIdentifier(workspaceUserEntity.getUserSchoolDataIdentifier().getDataSource(), workspaceUserEntity.getUserSchoolDataIdentifier().getIdentifier());
+      UserEntityProperty lastWorkspaces = userEntityController.getUserEntityPropertyByKey(userEntity, "last-workspaces");
+      
+      if (lastWorkspaces != null) {
+        if (StringUtils.isEmpty(lastWorkspaces.getValue())) {
```
Comment (pkukkon) [inline muikku-core/src/main/java/fi/otavanopisto/muikku/users/WorkspaceUserEntityController.java:224]: Should most likely be `if (!StringUtils.isEmpty(lastWorkspaces.getValue()))`

File: muikku-core/src/main/java/fi/otavanopisto/muikku/users/WorkspaceUserEntityController.java [hunk 3]
```diff
@@ -205,6 +215,38 @@ public WorkspaceUserEntity updateIdentifier(WorkspaceUserEntity workspaceUserEnt
...
+          catch (JsonProcessingException e) {
+            logger.log(Level.WARNING, String.format("Parsing last workspaces of user %d failed: %s", userEntity.getId(), e.getMessage()));
+          }
+          
+          if(lastWorkspaceList != null) {
+            Long workspaceEntityId = workspaceUserEntity.getWorkspaceEntity().getId();
+            lastWorkspaceList.removeIf(item -> workspaceEntityId.equals(item.getWorkspaceId()));
+          }
+          
+          try {
```
Comment (pkukkon) [inline muikku-core/src/main/java/fi/otavanopisto/muikku/users/WorkspaceUserEntityController.java:241]: Might as well compare list size before and after modification; if it's the same, no need to update the property as nothing has changed.

**Review by pkukkon — CHANGES_REQUESTED** [review]

File: muikku-core/src/main/java/fi/otavanopisto/muikku/users/WorkspaceUserEntityController.java (hunk 3 above)
Comment (aviljakaine) [reply muikku-core/src/main/java/fi/otavanopisto/muikku/users/WorkspaceUserEntityController.java:241]: Also, if lastWorkspaceList was null (the previous if block), is there any need to save?

These all could probably be combined under one try-catch too as they are depending on each previous step in the sequence.

File: muikku-core/src/main/java/fi/otavanopisto/muikku/users/WorkspaceUserEntityController.java (hunk 2 above)
Comment (aviljakaine) [reply muikku-core/src/main/java/fi/otavanopisto/muikku/users/WorkspaceUserEntityController.java:224]: Or `StringUtils.isNotEmpty(...`

**Review by pkukkon — APPROVED** [review]

**Pull request merged**